File: src/benchmarks/mnist/benchmark.ts
import { Model } from "../../interfaces.ts";
import { MNISTDataloader } from "./dataloader.ts";
import { TFNeuralNet } from "../../models/tf-neural-net/index.ts";
import * as tf from "npm:@tensorflow/tfjs";

export class MNISTBenchmark {
  private DATA_FOLDER = "src/benchmarks/mnist/data/";
  public data = new MNISTDataloader(
    this.DATA_FOLDER + "train/images",
    this.DATA_FOLDER + "train/labels",
    this.DATA_FOLDER + "test/images",
    this.DATA_FOLDER + "test/labels"
  ).loadAll();

  constructor(private model: Model | TFNeuralNet) {}

  private normalizeImage(image: number[][]) {
    return image.flat().map((p) => p / 255);
  }

  private normalizeLabel(label: number): number[] {
    const expectedOutput = Array(this.model.params.outputNodes).fill(0);
    expectedOutput[label] = 1;
    return expectedOutput;
  }

  public async train(epochCount: number) {
    const { images, labels } = (await this.data).train;

    // TFJS
    const batchSize = 100;
    if (this.model instanceof TFNeuralNet) {
      for (let i = 0; i <= 10000; i += batchSize) {
        const batchImages = images.slice(i, i + batchSize);
        const batchLabels = labels.slice(i, i + batchSize);
        console.log(`${i + 1} of ${images.length}`);

        // normalize data
        const input = tf.tidy(() => tf.tensor2d(batchImages.map(this.normalizeImage)));
        const output = tf.tidy(() => tf.tensor2d(batchLabels.map((label) => this.normalizeLabel(label))));

        await this.model.model.fit(input, output, {
          epochs: epochCount,
          batchSize,
        });

        // Dispose tensors to free memory
        input.dispose();
        output.dispose();
      }
      return;
    }

    // Model
    for (let epoch = 1; epoch <= epochCount; epoch++) {
      for (let i = 0; i < images.length; i++) {
        console.log(`${i + 1} of ${images.length}`);
        const image = images[i];
        const label = labels[i];

        // normalize data
        const input = this.normalizeImage(image);
        const output = this.normalizeLabel(label);

        await this.model.train(input, output);
      }
      console.log(`Epoch ${epoch} complete.`);
    }
  }

  public async test(): Promise<number> {
    const { images, labels } = (await this.data).test;

    // TFJS
    if (this.model instanceof TFNeuralNet) {
      // normalize data
      const input = images.map(this.normalizeImage);
      const output = labels.map((label) => this.normalizeLabel(label));

      const result = await this.model.model.evaluate(tf.tensor2d(input), tf.tensor2d(output));
      return Array.isArray(result) ? (await result[1].data())[0] : (await result.dataSync())[0];
    }

    // Models
    let correctCount = 0;

    for (let i = 0; i < images.length; i++) {
      const image = images[i];
      const label = labels[i];

      // normalize data
      const input = this.normalizeImage(image);
      const output = await this.model.feedForward(input);
      const predicted = output.indexOf(Math.max(...output));
      // console.log(predicted, label);

      if (predicted === label) correctCount++;
    }

    return correctCount / images.length;
  }
}

File: src/benchmarks/mnist/benchmark.rs
use std::path::PathBuf;

use crate::models::SimpleNeuralNet;

use super::MNISTDataloader;

pub struct MNISTBenchmark {
    neural_net: SimpleNeuralNet,
    data_loader: MNISTDataloader,
}

impl MNISTBenchmark {
    pub fn new(neural_net: SimpleNeuralNet) -> Self {
        let data_folder = PathBuf::from("src/benchmarks/mnist/data/");
        let data_loader = MNISTDataloader::new(
            data_folder.join("train/images").to_str().unwrap().to_string(),
            data_folder.join("train/labels").to_str().unwrap().to_string(),
            data_folder.join("test/images").to_str().unwrap().to_string(),
            data_folder.join("test/labels").to_str().unwrap().to_string(),
        );

        MNISTBenchmark {
            neural_net,
            data_loader,
        }
    }

    fn normalize_image(&self, image: &[Vec<u8>]) -> Vec<f64> {
        image.iter().flatten().map(|&p| p as f64 / 255.0).collect()
    }

    fn normalize_label(&self, label: u8) -> Vec<f64> {
        let mut expected_output = vec![0.0; self.neural_net.output_nodes];
        expected_output[label as usize] = 1.0;
        expected_output
    }

    pub fn train(&mut self, epoch_count: usize) {
        if let Ok((train_data, _)) = self.data_loader.load_all() {
            let (train_images, train_labels) = train_data;
            for epoch in 1..=epoch_count {
                for (i, image) in train_images.iter().enumerate() {
                    println!("{i}");
                    let label = train_labels[i];
                    let input = self.normalize_image(image);
                    let output = self.normalize_label(label);

                    self.neural_net.train(&input, &output);
                }
                println!("Epoch {} complete.", epoch);
            }
        }
    }

    pub fn test(&self) -> f64 {
        if let Ok((_, test_data)) = self.data_loader.load_all() {
            let (test_images, test_labels) = test_data;
            let mut correct_count = 0;

            for (i, image) in test_images.iter().enumerate() {
                let label = test_labels[i];
                let input = self.normalize_image(image);
                let output = self.neural_net.feed_forward(&input);
                let predicted = output.iter().enumerate().max_by(|(_, a), (_, b)| a.partial_cmp(b).unwrap()).map(|(index, _)| index).unwrap();

                if predicted == label as usize {
                    correct_count += 1;
                }
            }

            correct_count as f64 / test_images.len() as f64
        } else {
            0.0
        }
    }
}
